flink-connector-pinot/src/main/java/org/apache/flink/streaming/connectors/pinot/writer/PinotSinkWriter.java

```diff
@@ -0,0 +1,153 @@
+     * {@link PinotSinkCommittable}s. If flush is not set, only currently non-active
+     * {@link PinotSinkCommittable}s are transformed into {@link PinotSinkCommittable}s.
+     * To convert a {@link PinotWriterSegment} into a {@link PinotSinkCommittable} the data gets
+     * written to the shared filesystem. Moreover, minimum and maximum timestamps are identified.
+     * Finally, all {@link PinotWriterSegment}s transformed into {@link PinotSinkCommittable}s are
+     * removed from {@link #activeSegments}.
+     *
+     * @param flush Flush all currently known elements into the {@link PinotSinkCommittable}s
+     * @return List of {@link PinotSinkCommittable} to process in {@link org.apache.flink.streaming.connectors.pinot.committer.PinotSinkGlobalCommitter}
+     * @throws IOException
+     */
+    @Override
+    public List<PinotSinkCommittable> prepareCommit(boolean flush) throws IOException {
+        // Identify segments to commit. If the flush argument is set all segments shall be committed.
+        // Otherwise, take only those PinotWriterSegments that do not accept any more elements.
+        List<PinotWriterSegment<IN>> segmentsToCommit = activeSegments.stream()
+                .filter(s -> flush || !s.acceptsElements())
+                .collect(Collectors.toList());
+        LOG.debug("Identified {} segments to commit [subtaskId={}]", segmentsToCommit.size(), subtaskId);
+
+        LOG.debug("Creating committables... [subtaskId={}]", subtaskId);
+        List<PinotSinkCommittable> committables = new ArrayList<>();
+        for (final PinotWriterSegment<IN> segment : segmentsToCommit) {
+            committables.add(segment.prepareCommit());
+        }
+        LOG.debug("Created {} committables [subtaskId={}]", committables.size(), subtaskId);
+
+        // Remove all PinotWriterSegments that will be emitted within the committables.
+        activeSegments.removeAll(segmentsToCommit);
+        return committables;
+    }
+
+    /**
+     * Gets the {@link PinotWriterSegment} still accepting elements or creates a new one.
+     *
+     * @return {@link PinotWriterSegment} accepting at least one more element
+     */
+    private PinotWriterSegment<IN> getOrCreateInProgressSegment() {
+        final PinotWriterSegment<IN> latestSegment = Iterables.getLast(activeSegments, null);
+        if (latestSegment == null || !latestSegment.acceptsElements()) {
+            final PinotWriterSegment<IN> inProgressSegment = new PinotWriterSegment<>(maxRowsPerSegment, jsonSerializer, fsAdapter);
+            activeSegments.add(inProgressSegment);
+            return inProgressSegment;
+        }
+        return latestSegment;
+    }
+
+    /**
+     * As we do not need to save any information in snapshots,
+     * this method always returns an empty ArrayList.
+     *
+     * @return always an empty ArrayList
+     */
+    @Override
+    public List<Void> snapshotState() {
```

Comment:
I would expect that we have to save the path to the active files on every checkpoint.

Reply:
Not sure why we would need to store anything here. Could you elaborate on what you consider to be 'active files' ?
From my understanding we cannot reuse previously written `PinotWriterSegment`s, as elements might get sent to different subTasks when recovering from failure.

Reply:
You are totally right, many thanks for pointing that out! I've included a `PinotWriterState` that snapshots the necessary information (elements, min&max timestamps) from the `PinotWriterSegment` which is currently in-progress.  
Upon `PinotSinkWriter` creation the state would then get recovered and transformed into a new in-progress `PinotWriterSegment`.